StreamChatClient_v3/Controllers/Controller.swift

```diff
@@ -6,10 +6,36 @@ import Foundation
 
 /// The base class for all controllers. Not meant to be used directly.
 public class Controller {
+    /// Describes the possible states of Controller
+    public enum State: Equatable {
+        /// The controller is idle. Call `startUpdating` to start listening for changes.
+        case idle
+        
+        /// The controller is active and is listening to changes.
+        case active
+    }
+    
+    /// The current state of the Controller.
+    public internal(set) var state: State = .idle
+    
     /// The queue which is used to perform callback calls. The default value is `.main`.
     public var callbackQueue: DispatchQueue = .main
 }
 
+extension Controller {
+    /// A helper function to ensure the callback is performed on the callback queue.
+    func callback(_ action: @escaping () -> Void) {
+        if callbackQueue == DispatchQueue.main, Thread.current.isMainThread {
+            // Perform the action on the main queue synchronously
+            action()
+        } else {
+            callbackQueue.async {
+                action()
+            }
+        }
```

Comment:
isn't calling `async` on the same queue safe? GCD will optimize it and will just call the block anyway, seems redundant

Reply:
Not sure about other queues, but it doesn't work like this for the main queue.

Reply:
fixed in a858a7b19b6301d6b160ae6dd07768d24831733e